builders/txDataBuilder.go

```diff
@@ -0,0 +1,184 @@
+	builder.function = function
+
+	return builder
+}
+
+// CallerAddress sets the caller address
+func (builder *txDataBuilder) CallerAddress(address core.AddressHandler) TxDataBuilder {
+	err := builder.checkAddress(address)
+	if err != nil {
+		builder.err = err
+		return builder
+	}
+
+	builder.callerAddr = address.AddressAsBech32String()
+
+	return builder
+}
+
+// Address sets the destination address
+func (builder *txDataBuilder) Address(address core.AddressHandler) TxDataBuilder {
+	err := builder.checkAddress(address)
+	if err != nil {
+		builder.err = err
+		return builder
+	}
+
+	builder.address = address.AddressAsBech32String()
+
+	return builder
+}
+
+// ArgHexString adds the provided hex string to the arguments list
+func (builder *txDataBuilder) ArgHexString(hexed string) TxDataBuilder {
+	_, err := hex.DecodeString(hexed)
+	if err != nil {
+		builder.err = fmt.Errorf("%w in builder.ArgHexString for string %s", err, hexed)
+		return builder
+	}
+
+	builder.args = append(builder.args, hexed)
+
+	return builder
+}
+
+// ArgAddress adds the provided address to the arguments list
+func (builder *txDataBuilder) ArgAddress(address core.AddressHandler) TxDataBuilder {
+	err := builder.checkAddress(address)
+	if err != nil {
+		builder.err = err
+		return builder
+	}
+
+	return builder.addBytes(address.AddressBytes())
+}
+
+func (builder *txDataBuilder) checkAddress(address core.AddressHandler) error {
+	if check.IfNil(address) {
+		return fmt.Errorf("%w in builder.checkAddress", ErrNilAddress)
+	}
+	if len(address.AddressBytes()) == 0 {
+		return fmt.Errorf("%w in builder.checkAddress", ErrInvalidAddress)
+	}
+
+	return nil
+}
+
+// ArgBigInt adds the provided value to the arguments list
+func (builder *txDataBuilder) ArgBigInt(value *big.Int) TxDataBuilder {
+	if value == nil {
+		builder.err = fmt.Errorf("%w in builder.ArgBigInt", ErrNilValue)
+		return builder
+	}
+
+	return builder.addBytes(value.Bytes())
+}
+
+// ArgInt64 adds the provided value to the arguments list
+func (builder *txDataBuilder) ArgInt64(value int64) TxDataBuilder {
+	b := big.NewInt(value)
+
+	return builder.addBytes(b.Bytes())
+}
+
+// ArgBytes adds the provided bytes to the arguments list. The parameter should contain at least one byte
+func (builder *txDataBuilder) ArgBytes(bytes []byte) TxDataBuilder {
+	if len(bytes) == 0 {
+		builder.err = fmt.Errorf("%w in builder.ArgBytes", ErrInvalidValue)
+	}
+
+	builder.args = append(builder.args, hex.EncodeToString(bytes))
+
+	return builder
+}
+
+func (builder *txDataBuilder) addBytes(bytes []byte) TxDataBuilder {
+	if len(bytes) == 0 {
+		bytes = []byte{0}
+	}
+
+	builder.args = append(builder.args, hex.EncodeToString(bytes))
+
+	return builder
+}
+
+// ToDataString returns the formatted data string ready to be used in a transaction call
+func (builder *txDataBuilder) ToDataString() (string, error) {
+	if builder.err != nil {
+		return "", builder.err
```

Comment:
interesting approach. 
Suggestion: error check should be on all exported functions and only propagate the error further if any. This way, as a caller, you'll receive the earliest error so you'll keep trying from the begging to the end of the process.

Example:
```
NewDataBuilder()
   .ArgHexString("6f6b") // all good here
   .ArgAddress(nil)  // will set the error here
   .ArgHexString("6b6f") // will ignore the call with an early exit (if builder.err != nil)
   .ArgBytes([]byte{0,1,2,3}) // will ignore the call with an early exit
   .ToDataString() // will return the error from the step when setting address
```

Reply:
done in the baseBuilder struct